### DISCUSSION TITLE
Use web structured data to import books

### DISCUSSION THREAD

**Discussion by Gildas-GH** [mateusz-bak/openreads#734] [Ideas]
Hi, I have an idea to import books in the app, as OpenLibrary is often missing books and editions. I'd like to use Structured Data Markup on e-shopping websites to import author, cover image, title, ISBN and sometimes (when supported by the website) page number, edition.

Structured Data Markup is used by Google and other search engines to provide "Products" pages and other "intelligent" results (News, Q&A, recipes... https://developers.google.com/search/docs/appearance/structured-data/search-gallery)

Learn more about structured data : https://developers.google.com/search/docs/appearance/structured-data/intro-structured-data

Most of online shops use this markup to make their SEO better. We could also use it to get book info. I found two ways to implement it : 
- User gives a link and we scrape it to find structured data
- User gives an ISBN and we find info on a pre-defined list of bookshops

A structured data parser already exists in Flutter : [metadata_fetch](https://pub.dev/documentation/metadata_fetch/latest/)

### Examples

- French website : https://validator.schema.org/#url=https%3A%2F%2Fwww.placedeslibraires.fr%2Flivre%2F9782266320481-dune-tome-1-frank-herbert%2F (description, title, author, image, isbn...)
- Slovenian website : https://validator.schema.org/#url=https%3A%2F%2Femka.si%2Fproducts%2Fpesceni-planet-zepna-1 (only name, image, isbn)
- Goodreads : https://validator.schema.org/#url=https%3A%2F%2Fwww.goodreads.com%2Fbook%2Fshow%2F44767458-dune
- German website : https://validator.schema.org/#url=https%3A%2F%2Fwww.kulturkaufhaus.de%2Fen%2Fdetail%2FISBN-9781473655324%2FFrank-Herbert%2FDune-Messiah%3Fbpmctrl%3Dbpmrownr.1%257Cforeign.516539-1-0-0
...

The proposed system is independent of the website the user wishes to use. It will work on most e-commerce sites in many countries. 

For now I tried to fix some bugs in Openreads (first time using Flutter) but I was not able to test my code on a device (even virtual). If I manage to get it working, I'll try to implement it myself, if you'd like to have this system on Openreads. Thanks !

**Comment by AWFeldick** [comment]
This is a cool idea that I'd use for sure. 

Note though, the package you linked ([metadata_fetch](https://pub.dev/packages/metadata_fetch)) isn't really actually a general structured data parser. It only collects a webpage's title, description, image, and url. I'm not actually finding a good structured data parser available in Dart either, so to implement this might require writing a parser that can detect book data from scratch.

Even so, I might take a stab at this this weekend for fun. Though I'm brand new to Dart, Flutter, and app development so it might not be pretty, lol.

**Comment by Gildas-GH** [comment]
See the different parsers, including JSON-LD: https://github.com/jg-l/metadata_fetch?tab=readme-ov-file#manually-specify-which-metadata-parser-to-use
